### PR TITLE
Use conda-build=3

### PR DESCRIPTION
**Pull request by isuruf**
cc @conda-forge/core 

- [x] Add c3i LICENSE
- [x] Restore the recipes/examples folder

### CODE REVIEW

**Comment by isuruf** [comment]
@msarahan, do you know why circleci works, but not appveyor?

**Comment by jakirkham** [comment]
Because the [`example`]( https://travis-ci.org/conda-forge/staged-recipes/builds/346627887#L230 ) gets removed from PRs. Same story with AppVeyor. This would happen on CircleCI; however, as you are running CircleCI on your fork, the `example` is not removed. If you disabled CircleCI on your fork and pushed a fresh commit, it would have the same issue.

This comes down to [this code]( https://github.com/conda-forge/staged-recipes/blob/ca1db5f80ffee72d43a0efc960746f15f026601b/.travis_scripts/build_all#L36-L42 ) or similar for the other CIs, which makes sure we don't build recipes in PRs that were merged into `master`. Expect if you commented that code or changed `example2` and `example3` to not rely on `example`, the problem would be resolved.

If you do merge these examples, would double check that [this code]( https://github.com/conda-forge/staged-recipes/blob/master/.travis_scripts/create_feedstocks.py#L38-L41 ) keeps them from going into feedstocks. Seems like it should though. Might also be worth moving all the example recipes into one big `example` directory to keep the line noise low.

**Comment by isuruf** [comment]
I've removed the example recipe for now. example2 now depends on example3, but example2 is built first. Is this a conda-build issue or was it never meant to replace conda-build-all?

**Comment by msarahan** [comment]
conda-build can build more than one recipe at a time, but it currently does not do topological sorting for which should be built first.  We'd like to add that eventually, but it's not high priority for us.  We use conda-concourse-ci internally for this purpose.

**Comment by jakirkham** [comment]
So I think the issue is `example` was moved to `example3`. Hence we are running into the same issue as before. Would restore `example` and just add a brand new copy of `example3`. Happy to push a commit to your PR if you would like assist.

**Comment by isuruf** [comment]
> conda-build can build more than one recipe at a time, but it currently does not do topological sorting for which should be built first.

Okay. I guess I can write a simple python script that uses `conda_concourse_ci.compute_build_graph.construct_graph`. @msarahan, I don't see `conda-concourse-ci` in defaults. Is it meant for internal use only?

**Comment by isuruf** [comment]
> So I think the issue is example was moved to example3.

I don't think that's an issue here, but feel free to push to this PR.

**Comment by isuruf** [comment]
Here's a script that uses `conda_concourse_ci`. It uses code only from https://github.com/conda/conda-concourse-ci/blob/master/conda_concourse_ci/compute_build_graph.py. We can even vendor it, since it's just one file. We'll need a new dependency, `networkx`.

```python
import conda_build.conda_interface
import networkx as nx
import conda_build.api
from conda_concourse_ci.compute_build_graph import construct_graph
import argparse
import os

def build_all(recipes_dir, arch, variant_config_files):
    index = conda_build.conda_interface.get_index(channel_urls=['conda-forge'])
    conda_resolve = conda_build.conda_interface.Resolve(index)

    config = conda_build.api.Config(variant_config_files=variant_config_files)

    worker={'platform': get_host_platform(), 'arch': arch, 'label': get_host_platform()}

    G = construct_graph(recipes_dir, worker=worker, run='build', conda_resolve=conda_resolve,
                        folders=os.listdir(recipes_dir), config=config)

    order = list(nx.topological_sort(G))
    order.reverse()

    print('Building packages')
    print('\n'.join(order))

    for node in order:
        conda_build.api.build(G.node[node]['meta'])


def get_host_platform():
    from sys import platform
    if platform == "linux" or platform == "linux2":
        return "linux"
    elif platform == "darwin":
        return "osx"
    elif platform == "win32":
        return "win"


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument('recipes_dir', default=os.getcwd(), help='Directory where the recipes are')
    parser.add_argument('--arch', default='64', help='target architecture (64 or 32)')
    parser.add_argument('-m','--variant-config-files', action='append', help='variant config files')
    args = parser.parse_args()
    build_all(args.recipes_dir, args.arch, args.variant_config_files)
```

File: .ci_support/build_all.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    config = conda_build.api.Config(variant_config_files=variant_config_files, arch=arch,
+                                    channel_urls=channel_urls)
+
+    worker={'platform': get_host_platform(), 'arch': arch, 'label': get_host_platform()}
+
+    G = construct_graph(recipes_dir, worker=worker, run='build', conda_resolve=conda_resolve,
+                        folders=os.listdir(recipes_dir), config=config)
+
+    order = list(nx.topological_sort(G))
+    order.reverse()
```
Comment (CJ-Wright) [inline .ci_support/build_all.py]: Why is the build order reversed? (wouldn't this build the bottom of the graph first?)

File: recipes/example3/meta.yaml [hunk 2]
```diff
@@ -10,7 +10,7 @@
 #  `conda install openssl -c conda-forge``
 
 package:
-  name: {{ name|lower }}
+  name: simplejson2
```
Comment (scopatz) [inline recipes/example3/meta.yaml]: Is this on purpose?

Comment (isuruf) [reply recipes/example3/meta.yaml]: Nope, I'll be reverting the examples once I know this works.

File: .ci_support/build_all.py (hunk 1 above)
Comment (msarahan) [reply .ci_support/build_all.py]: Could be that I build the graph backward when adding the nodes.  I have considered going back and reversing it, but haven't found time.  When I have tried, it ended up exploding in my face.

Comment (isuruf) [reply .ci_support/build_all.py]: It depends on how the edge direction is defined. reverse here gives the order that I need

**Comment by isuruf** [comment]
This passes on appveyor in my fork.
I can use some help in getting linux to work. Need to copy the python files in `.ci_support` to the docker container and use them.

**Comment by isuruf** [comment]
Strange errors in linux and osx. @msarahan, any suggestions?

**Comment by msarahan** [comment]
In both cases, it's something to do with the system compilers and the resulting binary search path:

https://travis-ci.org/conda-forge/staged-recipes/builds/346956154#L302

On circle:
```
Fixing permissions
patchelf: file: /home/conda/staged-recipes/build_artifacts/simplejson2_1519761708098/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold/lib/python2.7/site-packages/simplejson/_speedups.so
    setting rpath to: $ORIGIN/../../..
WARNING (simplejson2,_speedups.so): did not find - or even know where to look for: /lib64/libc.so.6
WARNING (simplejson2,_speedups.so): did not find - or even know where to look for: /lib64/libpthread.so.0
```

I don't know anything more off the top of my head.  I'll try to reproduce in a docker container tomorrow.  @mingwandroid might have more ideas than me.

**Comment by isuruf** [comment]
I can reproduce this locally on linux.

It tries to test a py2.7 package on a py3.6 environment. definitely something wrong with my script
```

The following NEW packages will be INSTALLED:

    ca-certificates: 2018.1.18-0  conda-forge                                      
    ncurses:         5.9-10       conda-forge                                      
    openssl:         1.0.2n-0     conda-forge                                      
    python:          3.6.4-0      conda-forge                                      
    readline:        7.0-0        conda-forge                                      
    simplejson2:     3.8.2-py27_0 file:///home/conda/staged-recipes/build_artifacts
    sqlite:          3.20.1-2     conda-forge                                      
    tk:              8.6.7-0      conda-forge                                      
    xz:              5.2.3-0      conda-forge                                      
    zlib:            1.2.11-0     conda-forge 
```

**Comment by isuruf** [comment]
Tests pass. Don't merge till I restore the example folder.

Can I get a review?

**Comment by msarahan** [comment]
What conda version are you using?  You should be aware of https://github.com/conda/conda-build/issues/2707.  This sounds potentially related.

TLDR: probably better to pin conda to 4.3.x for now.  Kale is working on a fix.

File: .appveyor.yml [hunk 3]
```diff
@@ -5,13 +5,12 @@ environment:
 
   # Workaround for https://github.com/conda/conda-build/issues/636
   PYTHONIOENCODING: "UTF-8"
+  CONDA_INSTALL_LOCN: "C:\\Miniconda36-x64"
```
Comment (msarahan) [inline .appveyor.yml]: better to leave this as Miniconda3-x64 - that's a symlink to the current miniconda3 install on appveyor.  It'll be better to keep up-to-date that way (fewer updates, potentially)

File: .ci_support/compute_build_graph.py [hunk 4]
```diff
@@ -0,0 +1,597 @@
+#!/usr/bin/env python
```
Comment (msarahan) [inline .ci_support/compute_build_graph.py]: This is pretty big to vendor.  It's probably OK, but it would be good to re-examine this vendoring decision in a few months and see if it's still a good idea, or if it would be better to add a dependency on c3i.

File: .circleci/run_docker_build.sh [hunk 5]
```diff
@@ -54,15 +55,15 @@ echo "$config" > ~/.condarc
 # A lock sometimes occurs with incomplete builds. The lock file is stored in build_artifacts.
 conda clean --lock
 
-conda update conda conda-build
-conda install conda-build-all
-conda install conda-forge-build-setup
+source activate root
```
Comment (msarahan) [inline .circleci/run_docker_build.sh]: root is superfluous here.  ``source activate`` with no argument will activate root.  I prefer it also because conda 4.4 calls the root env "base," although referring to it as "root" still works, too.

File: .travis_scripts/create_feedstocks [hunk 6]
```diff
@@ -35,9 +35,7 @@ bash ~/miniconda.sh -b -p ~/miniconda
 )
 source ~/miniconda/bin/activate root
 
-conda install --yes --quiet git=2.12.2
-conda install -n root --yes --quiet jinja2 anaconda-client
-conda install --yes --quiet conda-build=2 conda-smithy
+conda install --yes --quiet conda-forge-ci-setup=1.* conda-smithy=3.* conda-forge-pinning git=2.12.2
```
Comment (msarahan) [inline .travis_scripts/create_feedstocks]: why pin git?

File: .ci_support/compute_build_graph.py (hunk 4 above)
Comment (isuruf) [reply .ci_support/compute_build_graph.py]: I prefer adding c3i as a dependency, but c3i is not on defaults, so I thought it was meant for internal use only.

File: recipes/example2/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+  # By default, the package will be built for the Python versions supported by conda-forge and for all major OSs.
+  # Add the line "skip: True  # [py<35]" (for example) to limit to Python 3.5 and newer, or "skip: True  # [not win]" to limit to Windows.
+  script: python -m pip install --no-deps --ignore-installed .
+
+requirements:
+  build:
+    - python
+    - pip
+    - simplejson2
+    # if your project compiles code (such as a C extension) then add `toolchain` as a build requirement.
```
Comment (msarahan) [inline recipes/example2/meta.yaml]: simplejson really should be built with its C extension.  Maybe this doesn't make the best example recipe?

File: .travis_scripts/create_feedstocks (hunk 6 above)
Comment (isuruf) [reply .travis_scripts/create_feedstocks]: Some of the scripts here fails with the newer git versions, https://github.com/conda-forge/staged-recipes/issues/3365

File: .ci_support/compute_build_graph.py (hunk 4 above)
Comment (msarahan) [reply .ci_support/compute_build_graph.py]: There's also a lot of unused stuff here.  I'm not sure if you want to remove it, or just leave it in for easier diffs down the line, should there be changes to both c3i and this file.

Comment (msarahan) [reply .ci_support/compute_build_graph.py]: We can work on getting it onto defaults.  It has only seen internal use, but I wouldn't say it's only meant for internal use.  It's just much more oriented towards our needs, but could grow more general if there's community use and input (PRs definitely welcome).

**Comment by mingwandroid** [comment]
```
WARNING (simplejson2,_speedups.so): did not find - or even know where to look for: /lib64/libc.so.6
```

Are you still seeing this? @msarahan we need to roll out a CB with my new PR I think.

@isuruf, you should turn those warnings into fatal errors (once you have a new CB) via:

```
echo "conda_build:"                                 >> ${CONDA_ROOT}/.condarc
echo "  error_overlinking: true"                    >> ${CONDA_ROOT}/.condarc
```

**Comment by isuruf** [comment]
> Are you still seeing this?

Yes.

> @isuruf, you should turn those warnings into fatal errors (once you have a new CB) via:

Would they give warnings if the new compiler packages aren't used? Also is there an equivalent `conda config` command instead of manually editing `.condarc`?

File: .ci_support/build_all.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+    print('Computed that there are {} distributions to build from {} recipes'.format(len(order), len(folders)))
+    print("Resolved dependencies, will be built in the following order:")
+    print('    '+'\n    '.join(order))
+
+    for node in order:
+        conda_build.api.build(G.node[node]['meta'])
```
Comment (isuruf) [inline .ci_support/build_all.py]: @msarahan, for building recipes with multiple outputs, what should I do? Get the recipe paths from the meta data above, put it into a OrderedSet and pass the path to `conda_build.api.build`

Comment (msarahan) [reply .ci_support/build_all.py]: that's basically correct.  It's the collapsing that happens in https://github.com/conda/conda-concourse-ci/blob/master/conda_concourse_ci/compute_build_graph.py#L274

It's not perfect, mind you.  There are still some dependency relationships that are not properly handled.  See https://github.com/conda/conda-concourse-ci/issues/69

It would be worth giving your proposal a try, to see how well it works.  I suspect that there might be some crummy corner cases that mess things up, but your idea should work most of the time.

**Comment by isuruf** [comment]
I've implemented building recipes using the meta path. It's inefficient because the recipes are rendered twice. Would passing the metadata objects after `collapse_subpackage_nodes` to conda_build work?

File: .ci_support/compute_build_graph.py (hunk 4 above)
Comment (jakirkham) [reply .ci_support/compute_build_graph.py]: Sorry to be late. What was the outcome on this?

File: .travis_scripts/create_feedstocks (hunk 6 above)
Comment (jakirkham) [reply .travis_scripts/create_feedstocks]: Yeah, this is a bit annoying, but I don't think it needs to be solved in this PR.

File: recipes/example2/meta.yaml (hunk 7 above)
Comment (jakirkham) [reply recipes/example2/meta.yaml]: I think it does, but then we don't use the `toolchain` in the example. Not sure why. 😕

File: .ci_support/compute_build_graph.py (hunk 4 above)
Comment (isuruf) [reply .ci_support/compute_build_graph.py]: https://github.com/conda/conda-concourse-ci/issues/71#issuecomment-369421274

**Comment by jakirkham** [comment]
> > @isuruf, you should turn those warnings into fatal errors (once you have a new CB) via:
>
> Would they give warnings if the new compiler packages aren't used? Also is there an equivalent `conda config` command instead of manually editing `.condarc`?

We should move this to a new [conda-forge-ci-setup]( https://github.com/conda-forge/conda-forge-ci-setup-feedstock ) issue.

File: .ci_support/compute_build_graph.py (hunk 4 above)
Comment (msarahan) [reply .ci_support/compute_build_graph.py]: This functionality will be split from c3i and make its way into conda-build.  No timeline right now, but hopefully soon.  I don't think it will take too much work to have it simply work.  Having it go fast will take more time.

**Comment by jakirkham** [comment]
What's the status on this? I know there were various discussions (some of which occurred on `c3i`), but haven't been following them closely.

**Comment by jakirkham** [comment]
Thoughts? :)

**Comment by scopatz** [comment]
Maybe we should pull the trigger while here at AnacondaCon

**Comment by msarahan** [comment]
definitely easier to coordinate responses to issues that way

**Comment by isuruf** [comment]
This builds recipes on staged-recipes, but if they used conda-build=3 features, it will stop working in the feedstock without the conda-smithy=3 update

**Comment by ocefpaf** [comment]
@isuruf let's try to merge this! Can you fix the conflicts? If not I'm happy to do it.

**Comment by isuruf** [comment]
> That seems like the current situation, correct? So no worse. :)

No. Currently, conda-build=3 features won't work in staged-recipes and feedstocks. When you merge this PR it will work in staged-recipes, but not in feedstocks which can be confusing.

**Comment by jakirkham** [comment]
Sure, what I meant was feedstocks will be just as broken on `conda-build` 3 as they are now.

Is the concern that recipes requiring `conda-build` 3 get merged before we release `conda-smithy` 3? How would you recommend we address this? Documentation? Something else?

**Comment by isuruf** [comment]
> Is the concern that recipes requiring conda-build 3 get merged before we release conda-smithy 3? 

Yes.

> How would you recommend we address this? Documentation? Something else?

Release conda-smithy=3 first since anything working on cb2 works with cb3, there will be no issues.

Merge this PR after conda-smithy=3 is released, then recipes with cb3 features will work in both staged-recipes and feedstocks

**Comment by jakirkham** [comment]
> Merge this PR after conda-smithy=3 is released, then recipes with cb3 features will work in both staged-recipes and feedstocks

So this gets at my larger question. Are we generally ok with merging this as is (aside from merge conflicts and order of merges)?

**Comment by scopatz** [comment]
@jakirkham - yes, I think everyone I talked to yesterday at the con about his @ocefpaf @CJ-Wright @ericdill @jjhelmus etc are OK with it

**Comment by mwcraig** [comment]
Not there, but I'm in favor of this moving ahead....

**Comment by scopatz** [comment]
Hmm I resolved the conflicts but it looks like it failed...

File: .appveyor.yml [hunk 9]
```diff
@@ -3,7 +3,7 @@ skip_commits:
 
 environment:
 
-  CONDA_INSTALL_LOCN: "C:\\Miniconda36-x64"
+  CONDA_INSTALL_LOCN: "C:\\Miniconda3-x64"
```
Comment (isuruf) [inline .appveyor.yml]: Please don't. https://github.com/conda/conda/issues/7144

Comment (msarahan) [reply .appveyor.yml]: Hmmm... OK.  I was probably chasing a red herring with the conda exe thing.  I will revert this.  Thanks for posting the issue.

Comment (jakirkham) [reply .appveyor.yml]: Given recent issues this seems to solve, are we ok leaving `Miniconda36` here short term? This may already be agreeable at this stage. Just trying to confirm.

**Comment by scopatz** [comment]
Pulling the trigger.

File: .travis_scripts/create_feedstocks (hunk 6 above)
Comment (isuruf) [reply .travis_scripts/create_feedstocks]: This was supposed to be merged after conda-smithy=3 is released. Now feedstock conversion is broken due to this

**Comment by ocefpaf** [comment]
Let's release conda-smith ;-)
I'll get to it soln if no one beats me to it.

**Comment by scopatz** [comment]
I agree @ocefpaf :)

**Comment by ocefpaf** [comment]
Let's change the milestone. I don't think those last two are crucial to our
workflow.

**Comment by isuruf** [comment]
Please test before releasing. I'd urge everyone from core to rerender at least 2 of their favourite recipes and see if it works.

**Comment by ocefpaf** [comment]
On it. Thanks @isuruf!

**Comment by scopatz** [comment]
I have tried rerendering a few of the harder packages, such as cyclus, and everything works like a dream

**Comment by isuruf** [comment]
Thanks @scopatz. Did you also try on CI?

**Comment by scopatz** [comment]
Not yet.

**Comment by isuruf** [comment]
Please have a look at https://github.com/conda-forge/conda-smithy/wiki/Release-Notes-3.0.0.rc1 and update if you find any mistakes

**Comment by scopatz** [comment]
Seems to work on cyclus and pyembree

* https://github.com/conda-forge/cyclus-feedstock/pull/31
* https://github.com/conda-forge/pyembree-feedstock/pull/11

File: .travis_scripts/create_feedstocks (hunk 6 above)
Comment (jakirkham) [reply .travis_scripts/create_feedstocks]: Added a simple reversion in PR ( https://github.com/conda-forge/staged-recipes/pull/5657 ) for now.
